### ISSUE TITLE
AutoEncoder with MAP Elites

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#642]
Technically, this issue is already part of #615 , but it is complex enough that we should separate it out. Let's let the current MAP-Elites in Picbreeder issue focus only on the fitness function from now on, and move all discussion of the AutoEncoder to here.

Note that you should only pursue this issue once you complete #635 . Basically, Issue #635 will help you set up a prototype to use here. The point of this particular issue is to focus on the low-level details of getting the autoencoder to work with MAP-Elites and Picbreeder.

To facilitate doing that, I have various hints directly from Adam. Adam trained his AutoEncoder in MATLAB. Here is a list of some parameters he used:
```
'MaxEpochs',500, ... 
'L2WeightRegularization',0.004, ... 
'SparsityRegularization',4, ... 
'SparsityProportion',0.15, ... 
```

I'm not entirely sure if all of these apply to us. Max epochs definitely has a corresponding parameter in DL4J, and so should the L2WeightRegularization. I'm not sure about the other two. However, the good news is that Adam basically used a standard AutoEncoder for MNIST and adapted it for Picbreeder by scaling the CPPN images down to 28 by 28 pixels (size of MNIST digits). In other words, he started from an MNIST autoencoder the same way we are.

Since you will only be looking at this issue after you finish #635 , the only focus left is to figure out how the AutoEncoder interacts with MAP Elites. Here is what Adam had to say about the steps:

1. After every N evaluations (I think for the paper I made it quite long like 2^18 evaluations to try to get things really converged) I retrain the AE.
2. I 'dump' the map, and reevaluate ever solution for new BC coordinates (which should change)
3. Refill the map using all of the previous solutions as one large 'batch' which fills in the map

N = 2^18 seems like a lot to me ... we may not have the computing power, and I think that this may explain why our prelim results are so simple. In any case, this will be a parameter that we can change. Dumping the map but then re-binning the solutions makes sense, because the previous autoencoder scores would be meaningless after retraining.

**Comment by schrum2** [comment]
I set up periodic saving of the whole archive every 10,000 individuals. This is less than the 2^18 proposed, which is more iterations than we even bother to run.

Note that this issue no longer relies on #635 since we can use a Python autoencoder #655 instead of the DL4J one in Java. We can make some progress on this issue now.

**Comment by schrum2** [comment]
The next step in this is to create a `GaierAutoencoderPictureBinningScheme` that includes citation information for this paper, https://dl.acm.org/doi/10.1145/3319619.3321897 , in the class javadoc comment at the top of the file.

This binning scheme has two dimensions:
1) Number of neurons in the network
2) Autoencoder's reconstruction loss (discretized).

The number of neurons is just like what we already have in a different binning scheme. The autoencoder loss is new. 

However, we don't want to reinitialize the AutoEncoder Python process for every image. We want to boot it up once and have it keep running in the background, waiting for inputs.

In `AutoEncoderProcess` add a static variable `currentProcess` that is of type `AutoEncoderProcess`. Add a method `getAutoEncoderProcess` that simply returns the value of `currentProcess`. Then, add a static method `resetAutoEncoder` that takes a String path to a pth file, kills the old `currentProcess` and replaces it with a new process based on the given pth file path. Add a static method `getReconstructionLoss` that takes a `BufferedImage` and sends it to the `currentProcess` and then returns the double value of the loss. 

We still need to talk about how to discretize the reconstruction loss, but start by setting up the code to access it

**Comment by annazwicker** [comment]
As it is, the autoencoder takes is able to take in pictures from parentDir, our personal training set for the autoencoder, and replicates them fairly well.  The autoencoder can replicate simple images (tested with just a simple dot image) with reasonable success.  However, when tested with the actual skull image, it fails to produce good results, despite the fitness being relatively high.

I believe the next step would be getting the autoencoder to somewhat successfully replicate the actual skull image (our end goal).  Perhaps testing increasingly complicated images would help to see exactly when an image becomes "too complicated" for the autoencoder to successfully reproduce.  I think something we could try doing is categorizing images and creating levels to test.  Each level would have several images that vary in how they look, but are roughly the same in complexity.  This might be a different way to determine at what complexity the autoencoder fails, but by using different images, what exactly it struggles to reproduce (certain shapes, contrasts, etc.).

**Comment by schrum2** [comment]
This last comment misunderstands the purpose of the autoencoder a bit. If you train the autoencoder on the images in the `parentDir` then you would not expect it to be able to successfully reconstruct the skull, and it would be shocking if it could, because nothing resembling a skull is in that training set. The fact that we cannot reconstruct the skull after training on that data set is a GOOD thing, because it shows that the skull would be considered "novel" (due to the high reconstruction loss) with respect to those images which do not look like the skull.

We don't want the autoencoder to produce a skull, we want a CPPN to produce a skull. Our MAP-Elites archive stores CPPNs that produce various images. This issue is concerned with changing how we decide what is worth keeping in the archive. So far, we have mainly used neuron count and link count, but we want to use the autoencoder reconstruction loss. What is supposed to happen is that we start with a bunch of images that look nothing like the skull, and train an autoencoder to reproduce those images. The autoencoder should succeed, meaning that the reconstruction loss for all images ALREADY in the archive would be low. This means that we would not bother keeping a lot of images like what we have already seen. However, if we produce some images with higher reconstruction loss, that means the image would go in a different bin, and would get a chance to survive until later, even if it doesn't look much like the skull either.

However, since we will periodically retrain the autoencoder on the archive and reclassify everything, we should gradually categorize a wider and wider variety of images as not being novel, so that the only way to succeed it to eventually reproduce the skull.

**Comment by schrum2** [comment]
We have something running which is a partial solution, and it seems to find better results much quicker than the CPPN complexity approach. However, I know that the process is incomplete, and I think it might be leading us to get stuck without progress. Currently, the code periodically trains an AutoEncoder on the archive, and this AutoEncoder is used to compute the reconstruction loss, which is one part of the behavior characteristic for the binning scheme.

However, every time the autoencoder gets retrained, we are just leaving the old images in the archive. Retraining the autoencoder should invalidate the reconstruction loss scores of all images present. Therefore, right after the autoencoder is retrained, we need to make a fresh empty archive and add the images to that with their reconstruction loss reevaluated with the new autoencoder. I'll mark a place in the code for you to carry this out.

**Comment by schrum2** [comment]
Actually, I did the archive reorganization code myself. Now running to test.

**Comment by schrum2** [comment]
Complete autoencoder-based binning is running and seems to be working. So far, has achieved the highest fitness we have seen. Here are the latest examples:

![0 8100758538167039-Neurons 48 loss 0 9,1 0](https://user-images.githubusercontent.com/8397236/122065696-c3a4d180-cdb7-11eb-8ea0-36252adc8486.jpg)
![0 8094002713727038-Neurons 38 loss 0 9,1 0](https://user-images.githubusercontent.com/8397236/122065715-c9021c00-cdb7-11eb-91df-254da626a2ce.jpg)
![0 7944382018139559-Neurons 13 loss 0 9,1 0](https://user-images.githubusercontent.com/8397236/122065799-d8816500-cdb7-11eb-92e1-7dc8043b83f4.jpg)

**Comment by schrum2** [comment]
There actually does seem to be a problem with the results so far. Although great progress has been made, it has stalled in a way that indicates a problem. Specifically, every image in the recent archives has a reconstruction loss index associated with [0.9,1.0]. However, all of the images look fairly similar. If all of the images look this similar, then the loss should be low. Need to see how the saved autoencoders perform and track what the problem is. Maybe they need to train for longer on more complex images?

**Comment by schrum2** [comment]
Every single saved image at all iterations is in the loss index for [0.9,1.0], so there is probably a problem with the binning scheme rather than the autoencoder. In fact, I ran the saved autoencoders on some images and got loss values much smaller than this, which is to be expected.

**Comment by annazwicker** [comment]
Conducting two fresh runs with the new changes.  Both using the autoencoder, one using the EnhancedCPPNPictureGenotype and the other using the regular genotype.

**Comment by schrum2** [comment]
We have these runs configured to go for a very long time, so the experiment is on going, but as of iteration 4650000 I am getting the following when using the Autoencoder binning with the plain CPPN genotypes:

![0 779661054263674-Neurons 30 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122233451-afc5a200-ce81-11eb-8a50-5fec22614a33.jpg)
![0 782874502686364-Neurons 22 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122233492-b94f0a00-ce81-11eb-8ffa-897b8cf5ac63.jpg)
![0 7847839454075047-Neurons 19 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122233582-cd930700-ce81-11eb-8253-e5060dd5f6b4.jpg)

**Comment by schrum2** [comment]
@annazwicker Do a comparison to the skull using the fitness troubleshooting tool. Post some of the comparison images

**Comment by annazwicker** [comment]
Here is the heat map generated by PictureTargetFitnessTroubleshoot using RMSE fitness.
![RMSESkullheatMap](https://user-images.githubusercontent.com/82399865/122240195-32049500-ce87-11eb-8028-bf9cc3e108fb.PNG)

**Comment by schrum2** [comment]
![0 794122735371648-Neurons 36 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122605244-ccf2a000-d03c-11eb-96af-3026c42fd57e.jpg)
![0 6612476176217363-Neurons 30 loss 0 2,0 3](https://user-images.githubusercontent.com/8397236/122605246-ccf2a000-d03c-11eb-96cc-6047fcd91ec2.jpg)
![0 7595415211859046-Neurons 42 loss 0 1,0 2](https://user-images.githubusercontent.com/8397236/122605248-cd8b3680-d03c-11eb-92c0-822a0d2597f7.jpg)
![0 7619043624354807-Neurons 40 loss 0 1,0 2](https://user-images.githubusercontent.com/8397236/122605249-cd8b3680-d03c-11eb-915d-59ca8aa7c283.jpg)
![0 7854681841702311-Neurons 15 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122605250-cd8b3680-d03c-11eb-8331-4595c51264cc.jpg)
![0 7971974973193322-Neurons 50 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122605251-cd8b3680-d03c-11eb-9e70-4422f4fedb36.jpg)
![0 67413363830932-Neurons 50 loss 0 2,0 3](https://user-images.githubusercontent.com/8397236/122605252-cd8b3680-d03c-11eb-933e-bd3cefca0aee.jpg)
![0 75851622886099-Neurons 30 loss 0 1,0 2](https://user-images.githubusercontent.com/8397236/122605253-cd8b3680-d03c-11eb-95aa-2a6f0e35927c.jpg)
![0 759318090753988-Neurons 37 loss 0 1,0 2](https://user-images.githubusercontent.com/8397236/122605254-cd8b3680-d03c-11eb-871a-70883e69f51b.jpg)
![0 793736548042129-Neurons 23 loss 0 0,0 1](https://user-images.githubusercontent.com/8397236/122605255-ce23cd00-d03c-11eb-83b3-b691d3475cea.jpg)

Though there is still room for improvement on this, I'm going to say that our ability to produce images that look like this means this issue can be closed. For comparison, here are images from the original research we are comparing against.

![](https://files.slack.com/files-pri/T8FKEK536-F025DUKQM8T/image.png)
